src/Razor/src/Microsoft.CodeAnalysis.Razor.Workspaces/ProjectSystem/DefaultProjectSnapshotManager.cs

```diff
@@ -657,9 +479,359 @@ protected virtual void NotifyListeners(ProjectChangeEventArgs e)
+                }
+
+                var newEntry = Change(entry, action, documentState);
+                if (!ReferenceEquals(newEntry?.State, entry.State))
+                {
+                    oldSnapshot = entry.GetSnapshot();
+                    newSnapshot = newEntry?.GetSnapshot() ?? oldSnapshot;
+                    using (var writeLock = upgradeableLock.GetWriteLock())
+                    {
+                        if (newEntry is null)
+                        {
+                            _projects_needsLock.Remove(projectFilePath);
+                        }
+                        else
+                        {
+                            _projects_needsLock[projectFilePath] = newEntry;
+                        }
+
+                        if (action is OpenDocumentAction)
+                        {
+                            if (documentFilePath is null)
+                            {
+                                throw new ArgumentNullException(nameof(documentFilePath));
+                            }
+
+                            _openDocuments_needsLock.Add(documentFilePath);
+                        }
+                        else if (action is CloseDocumentAction)
+                        {
+                            if (documentFilePath is null)
+                            {
+                                throw new ArgumentNullException(nameof(documentFilePath));
+                            }
+
+                            _openDocuments_needsLock.Remove(documentFilePath);
+                        }
+                    }
+
+                    return true;
+                }
+            }
+        }
+
+        oldSnapshot = newSnapshot = null;
+        return false;
+    }
+
+    private static Entry? Change(Entry originalEntry, IUpdateProjectAction action, DocumentState? documentState)
+    {
+        switch (action)
+        {
+            case AddDocumentAction addAction:
+                return new Entry(originalEntry.State.WithAddedHostDocument(addAction.NewDocument, CreateTextAndVersionFunc(addAction.TextLoader)));
+
+            case RemoveDocumentAction removeAction:
+                return new Entry(originalEntry.State.WithRemovedHostDocument(removeAction.OriginalDocument));
+
+            case CloseDocumentAction closeAction:
+                {
+                    if (documentState is null)
+                    {
+                        throw new ArgumentNullException(nameof(documentState));
+                    }
+
+                    var state = originalEntry.State.WithChangedHostDocument(
+                        documentState.HostDocument,
+                        async () => await closeAction.TextLoader.LoadTextAndVersionAsync(LoadTextOptions, cancellationToken: default).ConfigureAwait(false));
+                    return new Entry(state);
+                }
+
+            case OpenDocumentAction openAction:
+                if (documentState is null)
```

Comment:
Consider moving this inside the block below to be consistent with the code above.